data_safe_haven/utility/__init__.py

```diff
@@ -1,17 +1,23 @@
+from .annotated_types import AzureLocation, AzureLongName, Guid
 from .directories import config_dir
 from .enums import DatabaseSystem, SoftwarePackageCategory
 from .file_reader import FileReader
 from .logger import LoggingSingleton, NonLoggingSingleton
 from .singleton import Singleton
 from .types import PathType
+from .yaml_serialisable_model import YAMLSerialisableModel
 
 __all__ = [
+    "AzureLocation",
+    "AzureLongName",
     "config_dir",
     "DatabaseSystem",
     "FileReader",
+    "Guid",
```

Comment:
I'd like to keep things organised in a way that *everything* that needs to be imported from outside the same folder can be imported as `from data_safe_haven.<folder name> import <something>`. We might need some thoughts about better ways to group things (I absolutely *do* agree that `utility` has the potential to become a messy misc folder if we're not careful).

Reply:
For example, I think it would make sense to put `annotated_types.py` and `types.py` into a directory called `types` so that we could use `from data_safe_haven.types import AzureLocation`. Probably better as its own PR though.

Reply:
And you want at most one level, like `data_safe_have.types` rather than `data_safe_haven.utility.types`?

I don't think I'm too bothered unless you have a very large number of levels. As long as it makes sense. I could see annotated types being part of utilities, or its own module.

Is there a risk from confusing types and annotated types? 😱

Reply:
Here's the essay/blog I was looking for earlier: https://benhoyt.com/writings/python-api-design/. Have a look and see whether you agree or disagree :).

Reply:
I scanned through the blog post. I think the advice is all good.

Particularly, how making a simple, flat API means you can easily avoid breaking the API even when restructuring.
In our case, it isn't _as big_ a deal as we don't expect anyone to use this package as an API.

I think we can take that approach then, but we have (or might have) some other problems we need to fix,

- modules/packages that are not descriptive (enough). What is a `utility.Guid`, what is a `utility.FileReader`. It would be better then to have a `data_safe_haven.types`, `data_safe_haven.annotated_types` modules/packages. (In the first instance, they can probably be modules to avoid a boilerplate `__init__.py`.
- Redundant packages where you have a directory with only `__init.py__` and `a.py`, and the `__init__.py` just imports and exposes everything from `a.py`

Reply:
I think we have two APIs actually.

1. The structure of commands under the `dsh` CLI app (which I think you very sensibly wanted to organise as something like `dsh <segment> <verb>`. This one could be structured like the DSH data model ("these things belong to the SRE...")

2. The internal API of connections between modules inside the project. This one could be structured like the Python data model ("these things are all subclasses of X").

Reply:
Issue [here](https://github.com/alan-turing-institute/data-safe-haven/issues/1846)